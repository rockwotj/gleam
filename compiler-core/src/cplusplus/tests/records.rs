use crate::assert_cpp;

#[test]
fn single_variant() {
    assert_cpp!(
        r#"
pub type Person { Person(name: String, age: Int) }
"#,
    );
}

#[test]
fn unnamed_fields() {
    assert_cpp!(
        r#"
pub type Ip { Ip(String) }
"#,
    );
}

#[test]
fn multiple_variants() {
    assert_cpp!(
        r#"
pub type User {
  LoggedIn(name: String)
  Guest
}
"#,
    );
}

#[test]
fn shared_fields() {
    assert_cpp!(
        r#"
pub type Animal {
  Cat(name: String, likes_milk: Bool)
  Dog(name: String, barks: Bool)
}
"#,
    );
}

#[test]
fn generic_single_variant() {
    assert_cpp!(
        r#"
pub type Box(inner_type) {
  Box(inner: inner_type)
}
"#,
    );
}

#[test]
fn generic_multiple_variant() {
    assert_cpp!(
        r#"
pub type Either(left_type, right_type) {
  Left(v: left_type)
  Right(v: right_type)
}
"#,
    );
}

#[test]
fn singleton() {
    assert_cpp!(
        r#"
pub type Void { Void }

fn noop() -> Void {
  let v = Void;
  v
}
"#,
    );
}

#[test]
fn constructor_fn() {
    assert_cpp!(
        r#"
pub type Person { Person(name: String) }

fn new(name: String) -> Person {
  let constructor_fn = Person;
  constructor_fn(name)
}
"#,
    );
}
