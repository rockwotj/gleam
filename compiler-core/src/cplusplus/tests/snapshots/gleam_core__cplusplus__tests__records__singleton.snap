---
source: compiler-core/src/cplusplus/tests/records.rs
expression: "\npub type Void { Void }\n\nfn noop() -> Void {\n  let v = Void;\n  v\n}\n"
---
#ifndef MY_PACKAGE_MY_MODULE_H_
#define MY_PACKAGE_MY_MODULE_H_

#include <gleam.h>

namespace my {
namespace module {

class Void;

gleam::Ref<::my::module::Void> noop();

class Void {
public:
  Void() {}
  
private:
  
};

} // namespace my
} // namespace module

#endif // MY_PACKAGE_MY_MODULE_H_

---
#include "module.h"

namespace my {
namespace module {

gleam::Ref<::my::module::Void> noop() {
  gleam::Ref<::my::module::Void> v = gleam::MakeRef<::my::module::Void>();
  return v;
};

} // namespace my
} // namespace module

